Use CSSTree for style transformations (new)

strarsis opened the pull request:
## Description
This PR adds the [CSSTree](https://github.com/csstree/csstree) parser for wrapping the theme styles into a styles wrapper.

## How has this been tested?
I tested this on my local WordPress site.

## Types of changes
- Because a custom CSS parser is currently used, parse errors can occur, as in https://github.com/WordPress/gutenberg/issues/21569.
- [CSSTree](https://github.com/csstree/csstree) is a native JavaScript CSS parser that is already used in other major projects like svgo and csso, hence it should handle all CSS fast and well.
- [CSSTree](https://github.com/csstree/csstree) can also generate sourcemaps and even reuse existing sourcemaps. With some API changes it should be possible to add sourcemap support in Gutenberg, which means that developers can still see from where the injected styles came from 
(PoCs here: https://codepen.io/strarsis/pen/RwWRqZv and also https://codesandbox.io/s/busy-williamson-r1frp). (Not implemented with this PR, but `CSSTree` adds the foundation for it).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

strarsis committed:
Use CSSTree for style transformations.

strarsis committed:
Add test for `url()` inside (valid) data URL.

strarsis committed:
Re-add DOM traversal test.

strarsis committed:
Remove old CSS parser.

strarsis commented:
@youknowriad, @noahtallen: OK, PR is ready. The original structure is preserved.

youknowriad commented on packages/block-editor/src/utils/transform-styles/transforms/url-rewrite.js:
```diff
@@ -1,141 +1,29 @@
-
-	return raw;
-}
+import { URL } from 'whatwg-url';
```
I know we've gone back  and forth about this kind of  module and we have a @wordpress/url package.  @aduth is this  correct, should we  replace with our own module instead?

youknowriad commented on packages/block-editor/src/utils/transform-styles/transforms/test/url-rewrite.js:
```diff
@@ -21,14 +21,6 @@ describe( 'URL rewrite', () => {
 		expect( output ).toMatchSnapshot();
 	} );
 
-	it( 'should not replace absolute paths', () => {
```
Why did we remove this test?

strarsis replied:
Yes, the test was removed because URLs that start with a slash (e.g. `/images/image.jpg`) are still relative URLs, just relative to the host and normalized (without `..`s). However, if there should be still an exception for host-relative URLs this can be added.

youknowriad replied:
So you're saying that rewriting them or not still results in the same URL. That's fine (though not rewriting seems safer), but let's keep the test.

strarsis replied:
Alright, the test can be readded. Though the URL would be rewritten from `/something/` to `https://example.com/something/` because then it becomes an absolute URL.

youknowriad replied:
that's probably fine but since the spec is sometimes very large, it seems safer to just not rewrite it if possible.

strarsis replied:
@youknow: OK, then I add a test for hose-relative URLs and just skip them.

strarsis replied:
Code and test added. 👍

youknowriad commented:
Can we get a quick sanity checks from theme authors  here @kjellr @jasmussen to ensure the editor styles are still working as intended in the editor since this is a big rewrite.

@strarsis did you check if there's any performance impact here (initial loading)  good or bad? (`nom run test-performance`)?

strarsis committed:
Add code and test to skip host-relative URLs (e.g. `/images/image.jpg` when rewriting URLs.

strarsis commented:
@youknowriad: Hm, the `test-performance` script doesn't want to run through on my system:
````
$ npm run test-performance

> gutenberg@7.9.1 test-performance [...]/gutenberg
> wp-scripts test-e2e --config packages/e2e-tests/jest.performance.config.js

Downloading Chromium r737027 - 123.7 Mb [====================] 100% 0.0s
Chromium (737027) downloaded to [...]/gutenberg/node_modules/puppeteer/.local-chromium/linux-737027
● Validation Error:

  Module @wordpress/jest-console in the setupFilesAfterEnv option was not found.
         <rootDir> is: [...]/gutenberg/packages/e2e-tests
[...]
npm ERR! gutenberg@7.9.1 test-performance: `wp-scripts test-e2e --config packages/e2e-tests/jest.performance.config.js`
````
The `@wordpress/jest-console` package seems to be there though.

jasmussen commented:
Hello! Thanks for the PR and thanks for the ping.

I'm usually able to check out branches in forks using this process:

```
git remote add strarsis git@github.com:strarsis/gutenberg.git
git fetch strarsis
git checkout strarsis/css-tree-transform-styles-2 -B css-tree-transform-styles-2
```

However I'm doing something wrong, as I'm not able to check out and compile this branch, sorry about that. Let me know if you spot my error.

To respond at a more high level though, I'm afraid I wasn't familiar with CSStree — what feedback are you looking for? Should the block editor and editor styles ideally look visually unchanged before and after this PR? Or does it change anything with regards to how editor styles are authored? What would be the impact if/when the block editor lands in an iframe and the DOM continues to be lightened until at one point it may be mostly identical to the frontend?

strarsis commented:
@jasmussen: An iframe-based approach to style namespacing hasn't been implemented yet.
For now, this PR should improve the style-wrapping using a CSS parser/AST library (CSSTree).
It also fixes CSS-specific issues like https://github.com/WordPress/gutenberg/issues/21569 and https://github.com/WordPress/gutenberg/issues/16408 where the currently used parser doesn't support some syntax. Until the new iframe paradigm has been worked out and implemented, fixing the parse issues would be a good step because it can be a blocker in theme development as the styles won't load at all in editor.

jasmussen commented:
Thanks for taking the time, and thanks again for the PR. I'm not sure why the branch checkout didn't work, but it seems to me like the best way to test this PR is to verify current editor styles look mostly (exactly?) the same before and after this branch, correct?

As far as iframes go, while it still might be a ways out, it also might be [closer than we think](https://github.com/WordPress/gutenberg/pull/21102). While it's important that the idea of the iframe PR landing should not stop forward momentum — it might _not_ land, after all — it also seems prudent to think of the iframe as a serious possibility, and avoid painting ourselves into a corner if we can. But it sounds like you're saying making this work with an iframed editor should not be a big refactor?

youknowriad commented:
>  it also seems prudent to think of the iframe as a serious possibility, and avoid painting ourselves into a corner if we can. But it sounds like you're saying making this work with an iframed editor should not be a big refactor?

I think we should be prudent about the iframe, even if we merge it at first, the PR  there showed that we can't get rid of the non-iframed version quickly since the iframe version breaks blocks that rely on the global "window/document" object. 

That said, If ever we go full iframe, this PR and the whole editor styles normalization won't be needed anymore (so just remove the code).

I think this PR is a good PR code wise, avoid the bundled dependency and use a more maintained library. I'd like to check  these:

 - Editor styles still look the same (like master)
 - Performance and bundle size. (for some reason the bundle size  bot is not triggering here)

strarsis commented:
@youknowriad: How can this be better tested? I also have issues getting the performance test running, even on master. But I use WSL 1 on Windows, there may be subtle differences.

youknowriad commented:
Yeah, can't speak about Windows, but once the theme styles are validated properly, I'll help with the performance test.

strarsis commented:
@youknowriad: Ideally there should be some theme styles that can be tested against. I could put in all the WordPress core theme styles (minified) like TwentyTwenty.

oxyc commented:
I couldn't get the tests running on macos either. Same error as you but it used to work a few months back.

strarsis commented:
@oxyc: I managed to get it running. You have to invoke some scripts first and then run puppeteer in headless mode with some extra options.
Prepare the workdir: `npm run build:packages`, `npm run wp-env start`

Add fields to puppeteer launch options:
`packages/scripts/config/puppeteer.config.js`:
```js
module.exports = {
        launch: {
                devtools: process.env.PUPPETEER_DEVTOOLS === 'true',
                headless: process.env.PUPPETEER_HEADLESS !== 'false',
                slowMo: parseInt( process.env.PUPPETEER_SLOWMO, 10 ) || 0,
                args: [
"--disable-gpu",
"--renderer",
"--no-sandbox",
"--no-service-autorun",
"--no-experiments",
"--no-default-browser-check",
"--disable-dev-shm-usage",
"--disable-setuid-sandbox",
"--no-first-run",
"--no-zygote",
"--single-process",
"--disable-extensions",
                ],
        },
};
````

Then run the performance test:
`PUPPETEER_HEADLESS=true npm run test-performance`

However, while the performance test runs now on master, it fails with:
````
● Performance › Loading, typing and selecting blocks                                                                                                                                                                                                                                                                                                                                                                                  No node found for selector: .block-editor-inserter__search-input
````
(`packages/e2e-tests/specs/performance/performance.test.js:75`)

Related issue: https://github.com/WordPress/gutenberg/issues/22038

oxyc commented:
I just tried to test this again but `npm run test-performance` on masters throws random error message, just like mentioned in https://github.com/WordPress/gutenberg/issues/22045 before.

This time:

>     No node found for selector: tr[data-slug="gutenberg-test-plugin-disables-the-css-animations"] .activate a

and

>     Evaluation failed: ReferenceError: wp is not defined

strarsis commented:
@oxyc: Hm, this error stems from puppeteer/chromium automated testing, when a node can't be found for a selector. But this is on the DOM itself, no CSS involved. The error will also cause the rest of the code failing to load, hence the wp related error afterwards. I think these are issues with the test and changed DOM structure - or a build script that hadn't run and updated the (committed) build artifacts that are used in the automated testing.

youknowriad commented:
@oxyc The error you're having is because you're not loading the e2e test plugin on your environment (they are in `packages/e2e-tests/plugins`

youknowriad commented:
also right now on  master you can do this to compare two branches `./bin/plugin/cli perf branch1 branch2`

youknowriad commented:
Sorry for letting this PR stale, I'm still interested in landing this, do you  think you can rebase it again?

strarsis commented:
@youknowriad: Sure! 🐱

strarsis committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into css-tree-transform-styles-2

youknowriad commented on packages/block-editor/src/utils/transform-styles/index.js:
```diff
@@ -1,19 +1,14 @@
- * WordPress dependencies
- */
-import { compose } from '@wordpress/compose';
+import { map, compose } from 'lodash';
```
I had an error here when testing. "compose" is on @wordpress/element

youknowriad reviewed (commented):
After fixing the compose error, I tried this PR and it seems the styles are not applied properly. I can tell because the styles of the buttons in 2019, for instance, are not correct (they should be blue and not that rounded)

oxyc commented on packages/block-editor/src/utils/transform-styles/index.js:
```diff
@@ -31,11 +26,7 @@ const transformStyles = ( styles, wrapperClassName = '' ) => {
-		}
-
-		return css;
+		return traverse( css, compose( transforms ) );
```
This shouldn't use compose since the result of the second transform gets called with the result of the first, it should just be fed the original arguments, eg:

```js
return traverse( css, ( ...args ) => {
  transforms.map( ( transform ) => transform( ...args ) );
} );
```

But that would change the old `traverse` api

oxyc commented:
Currently getting:

```
  ● Performance › Loading, typing and selecting blocks

    expect(jest.fn()).not.toHaveErrored(expected)
    Expected mock function not to be called but it was called with:
    ["Warning: React does not recognize the `%s` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `%s` instead. If you accidentally passed it from a parent component, remove it from the DOM element.%s isSelected isselected·
```

But I guess shouldnt actually have affected the results since they already ran?

If that's true, the initial performance test for Loading time:

| 3 runs | Average time to load|Average time to DOM content load
|-|----|---
|Before|10582ms|10479ms
|After|10495ms|10392ms

strarsis commented:
I am sorry that I hadn't really responded yet, I am very busy right now.
But from the performance data it looks that the PR would actually increase the performance (though even just slightly)?

oxyc commented:
> After fixing the compose error, I tried this PR and it seems the styles are not applied properly. I can tell because the styles of the buttons in 2019, for instance, are not correct (they should be blue and not that rounded)

Found the bug.

Before:

```css
.editor-styles-wrapper .wp-block-button:not(.is-style-outline) .wp-block-button__link
```

After

```css
.editor-styles-wrapper .wp-block-button:not(.editor-styles-wrapper .is-style-outline) .wp-block-button__link
```

oxyc commented:
Here's a fix for the above bug: https://github.com/oxyc/gutenberg/commit/6b746cb436f606776aad8ba1dfb3cf64737178c8

And here's a diff of twentyninenteens styles before and after

https://gist.github.com/oxyc/479a82cda75d92d0f30ba770c8fbc60f#file-after-patch

oxyc commented:
Tested it with own theme where I first encountered https://github.com/WordPress/gutenberg/issues/16408 and found another bug:

```after.diff
 @-webkit-keyframes background-zoom-sm {
-	0% {
+	.editor-styles-wrapper 0% {
```

Fix: https://github.com/oxyc/gutenberg/commit/1c1590b39247857459e33096697946da25c11c63

So with the following fixes and a fix for the merge artifacts I'd call it LGTM:
- https://github.com/oxyc/gutenberg/commit/c3e2c10052273bfc2806deeed29752f85ff409a4
- https://github.com/oxyc/gutenberg/commit/6b746cb436f606776aad8ba1dfb3cf64737178c8
- https://github.com/oxyc/gutenberg/commit/1c1590b39247857459e33096697946da25c11c63

strarsis commented:
@oxyc: Awesome, thanks! I got very little time currently, it is nice that you further looked into it.

strarsis commented:
Superseded by new PR on top of latest upstream master: https://github.com/WordPress/gutenberg/pull/25514

strarsis commented:
New editor styles transformation approach using [PostCSS](https://github.com/WordPress/gutenberg/pull/49521) is now merged (many thanks @zaguiini)!
